### PR TITLE
[pyright] 1.1.339 -> 1.1.349

### PR DESCRIPTION
**Pull request by smackesey**
Internal companion PR: https://github.com/dagster-io/dagster/pull/19482

- Bump pyright version
- Update pyright pins
- Fix many type errors
- Add `pyproject.toml` for `with_airflow` example. This matches our other `with_*` examples and resolves an issue pyright was having correctly including `with_airflow` in the `requirements_pinned.txt` environment specification.
- Change name of `dagster_buildkite` to `dagster-buildkite`.  This matches our other `dagster-*` packages and resolves an issue pyright was having correctly including `dagster-buildkite` in the `requirements_pinned.txt` environment specification.

This bump of pyright surfaced many new type errors due to some changes in what code gets analyzed. I have fixed the errors in this PR after consultation with various engs who best know each affected part of the code.

There was also a breaking change in pragma interpretation-- many subcategories were broken out of the `reportGeneralTypeIssues` category, necessitating updating of some pragmas that targeted `reportGeneralTypeIssues`.

One notable change is in the annotation setup for nesting resources in configurable resources. Configurable resource inherits a descriptor interface that defines type annotations when specifying nested resources. This is correctly interpreted by pyright but not by mypy. Prior to this PR, mypy was not complaining about it, but something in this PR caused mypy to start-- I suspect the addition of an annotation somewhere caused mypy to start analyzing this code when it previously wasn't doing so. After consultation with @benpankow, we decided to circumvent the problematic annotations by changing them to `Any` until mypy fixes the problem or a better solution can be found.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19482** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19482?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19482?utm_source=stack-comment).

File: pyright/master/requirements-pinned.txt [hunk 1]
```diff
@@ -519,76 +520,76 @@ tinycss2==1.2.1
...
-wcwidth==0.2.12
+wcwidth==0.2.13
 webcolors==1.13
 webencodings==0.5.1
 websocket-client==1.7.0
 websockets==12.0
 Werkzeug==2.2.3
 widgetsnbextension==4.0.9
--e examples/with_airflow
+with_airflow==0.0.0
```
Comment (jmsanders) [inline pyright/master/requirements-pinned.txt:584]: This feels odd?

**Review by jmsanders — CHANGES_REQUESTED** [review]

File: .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py [hunk 2]
```diff
@@ -110,7 +110,8 @@ def build_repo_wide_pyright_steps() -> List[CommandStep]:
             "pip install -e python_modules/dagster[pyright] -e python_modules/dagster-pipes",
             "make pyright",
         )
-        .on_test_image(AvailablePythonVersion.get_default())
+        # 3.11 (current default) can't support the entire huge pyright environment as of 2024-01-29
```
Comment (jmsanders) [inline .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py:113]: Do we have a list of what we're waiting on for support?

**Review by jmsanders — APPROVED** [review]

File: python_modules/dagster/dagster/_serdes/serdes.py [hunk 3]
```diff
@@ -52,17 +52,16 @@
 
 from .errors import DeserializationError, SerdesUsageError, SerializationError
 
+if TYPE_CHECKING:
+    # There is no actual class backing Dataclasses, _typeshed provides this
+    # protocol.
+    from _typeshed import DataclassInstance
+
```
Comment (smackesey) [inline python_modules/dagster/dagster/_serdes/serdes.py:59]: Replaces custom `DataclassProtocol`

File: python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_type_signatures.py [hunk 4]
```diff
@@ -137,7 +136,7 @@ def test_type_signatures_alias():
 from pydantic import Field
 
 class ResourceWithAlias(ConfigurableResource):
-    _schema: str = Field(alias="schema")
+    my_schema: str = Field(alias="schema")
```
Comment (smackesey) [inline python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_type_signatures.py:139]: this fixes an incidental name conflict

File: .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py (hunk 2 above)
Comment (smackesey) [reply .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py]: this turned out to be a red herring, 3.11 works

**Comment by smackesey** [comment]
This turned from a pins-only to a full update PR since @jmsanders reviewed, so requesting review from @alangenfeld.

File: python_modules/dagster/dagster/_config/pythonic_config/typing_utils.py [hunk 5]
```diff
@@ -176,12 +173,21 @@ class FooResource(Resource):
...
+    #     value: Union[Self, "PartialResource[Self]"]
+    # to:
+    #     value: Union[Any, "PartialResource[Any]"]
+    # This is because of a bug in mypy that is incorrectly interpreting the
+    # signature and can cause a false positive type error for users. This only
+    # started being detected in our test_type_signatures.py tests on 2024-02-02
+    # when some annotations elsewhere were added, likely causing mypy to
+    # analyze code it was previously skipping. The annotation should be
+    # reverted when the bug is fixed or another solution that surface as type
+    # errors for mypy users is found.
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_config/pythonic_config/typing_utils.py:190]: how did you find this mypy issue?

File: python_modules/dagster/dagster/_serdes/serdes.py [hunk 6]
```diff
@@ -998,7 +997,9 @@ def _unpack_object(val: dict, whitelist_map: WhitelistMap, context: UnpackContex
 
     if "__mapping_items__" in val:
         return {
-            _unpack_value(k, whitelist_map, context): _unpack_value(v, whitelist_map, context)
+            cast(str, _unpack_value(k, whitelist_map, context)): _unpack_value(
+                v, whitelist_map, context
+            )
             for k, v in val["__mapping_items__"]
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_serdes/serdes.py:1003]: I think this `str` cast is wrong, the key type type for this is non-scalar but still keyable https://github.com/dagster-io/dagster/pull/18057

File: python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_type_signatures.py (hunk 4 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_type_signatures.py:139]: maybe preserve `_` prefix

File: python_modules/dagster/dagster_tests/storage_tests/branching_io_manager_tests/utils.py [hunk 7]
```diff
@@ -86,14 +86,22 @@ def __init__(self):
 
     def handle_output(self, context: OutputContext, obj: Any):
         keys = self._keys_from_context(context)
-        for key in keys:
-            self.values[key] = obj
+        if keys is None:
+            self.values[None] = obj
```
Comment (alangenfeld) [inline python_modules/dagster/dagster_tests/storage_tests/branching_io_manager_tests/utils.py:90]: this pattern is odd to me - did you need it to get tests passing?

**Review by alangenfeld — APPROVED** [review]

File: python_modules/dagster/dagster/_config/pythonic_config/typing_utils.py (hunk 5 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_config/pythonic_config/typing_utils.py:190]: Some tests in `test_type_signatures.py` started failing. @benpankow and I investigated and it turned out that mypy and pyright had differing interpretations of the annotations.

File: python_modules/dagster/dagster_tests/storage_tests/branching_io_manager_tests/utils.py (hunk 7 above)
Comment (smackesey) [reply python_modules/dagster/dagster_tests/storage_tests/branching_io_manager_tests/utils.py:90]: It's a refactor of what was already happening, and it was the simplest way to get the type-checking to pass. Previously `_keys_from_context` would sometimes return `[None]` (IMO quite confusing) and we would still have `self.values[None]` set, only implicitly.

File: python_modules/dagster/dagster/_serdes/serdes.py (hunk 6 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_serdes/serdes.py]: I see, OK I've changed to cast to `Any`.

**Pull request merged**